### PR TITLE
METRON-1643: Create a REGEX_ROUTING field transformation

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
A relatively common pattern is to choose the value of a field based on if another field matches a set of regexes.  This can be done via stellar with match, but if the list of possible regexes are large, that can be unwieldy.  To ease that burden, we should have a REGEX_ROUTING field transformation.

## Manual Test Script
Ensure that `ZOOKEEPER`, `BROKERLIST` and `METRON_HOME` are set.
* Create a `dummy` topic
```
/usr/hdp/current/kafka-broker/bin/kafka-topics.sh --zookeeper $ZOOKEEPER --create --topic dummy --partitions 1 --replication-factor 1
```
* Open $METRON_HOME/config/zookeeper/parsers/dummy.json and save
```
{
  "parserClassName" : "org.apache.metron.parsers.csv.CSVParser"
  ,"sensorTopic" : "dummy"
  , "parserConfig": {
     "columns" : { "name": 0
                 , "first_field": 1
                 , "timestamp" : 2
                 }
   }
  ,"fieldTransformations" : [
    {
     "transformation" : "REGEX_SELECT"
    ,"input" :  "name"
    ,"output" :  "is_metron"
    ,"config" : {
      "true" : "^metron.*"
                }
    }
                           ]

}
```
*  Push configs via `$METRON_HOME/bin/zk_load_configs.sh --mode PUSH -z $ZOOKEEPER -i $METRON_HOME/config/zookeeper/`
* Stop the parsers in ambari
* Start the `dummy` parser
```
$METRON_HOME/bin/start_parser_topology.sh -k $BROKERLIST -z $ZOOKEEPER -s dummy
```
* Open `~/data.json` and save:
```
notmetron,f1,1
metron,f2,2
metron,f3,3
metron,f4,4
metron,f5,5
metron,f6,6
metron,f7,7
notmetroneither,f8,8
```
* Send data into the `dummy` topic via
```
cat data.json | /usr/hdp/current/kafka-broker/bin/kafka-console-producer.sh --broker-list $BROKERLIST --topic dummy
```
You should examine the `dummy` index in elasticsearch and ensure that messages with the `name` field of `metron` have a field called `is_metron` set to `true`.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-platform/metron-parsers/README.md [hunk 1]
```diff
@@ -337,6 +337,28 @@ The following config will rename the fields `old_field` and `different_old_field
                       ]
 }
 ```
+* `REGEX_ROUTING` : This transformation lets users set an output field to one of a set of possibilities based on matching regexes. This has functional overlap with Stellar match statements, but the syntax is more bearable for large sets of conditionals.
```
Comment (ottobackwards) [inline metron-platform/metron-parsers/README.md]: I don't think we need to editorialize so much about the match statements here.
"This transformation is useful when the number or conditions are large enough to make a stellar language match statement unwieldy" would be better.

**Comment by ottobackwards** [comment]
I am not sure ROUTING is a good name for this.  This is more like a SELECT.

File: metron-platform/metron-parsers/README.md (hunk 1 above)
Comment (cestella) [reply metron-platform/metron-parsers/README.md]: Yeah, that was wrong of me.  I had a bad night, when I wrote that.  My bad.  Will correct right away.  `match` is and continues to be awesome. :)

**Comment by ottobackwards** [comment]
+1 to SELECT
What you are saying is SET FIELD to X if SELECT.
It would be SWITCH if it were a different X per matching regex

**Comment by cestella** [comment]
@ottobackwards Makes sense; I made the change and percolated it to METRON-1644 aka #1084

**Comment by cestella** [comment]
Ok, I updated with manual testing plan and I think I've made the changes requested.  Let me know what you guys think.

**Comment by nickwallen** [comment]
Can you provide a "routing in Stellar" versus "routing with REGEX_ROUTING" example that highlights the benefits of this approach?  

This go against our current trajectory of "do everything in Stellar" and I want to make sure I understand the benefits.

**Comment by cestella** [comment]
Sure, so the difference in the parser chaining example would be between the following
# Stellar
```
"fieldTransformations" : [
    {
     "transformation" : "STELLAR"
    ,"input" :  "pix_type"
    ,"output" :  "logical_source_type"
    ,"config" : {
      "logical_source_type" : "match{REGEXP_MATCH(pix_type, '^6-302.*') => 'cisco-6-302', REGEXP_MATCH(pix_type, '^5-304.*') => 'cisco-5-304', default => NULL}",
                }
    }
                           ]
```

# `REGEX_SELECT`
```
"fieldTransformations" : [
    {
     "transformation" : "REGEX_SELECT"
    ,"input" :  "pix_type"
    ,"output" :  "logical_source_type"
    ,"config" : {
      "cisco-6-302" : "^6-302.*",
      "cisco-5-304" : "^5-304.*"
                }
    }
                           ]
```

It bears mentioning that things get worse in stellar when:
1. You want to match on one of a set of regexs (e.g. `match { REGEXP_MATCH(pix_type, 'regex1') or REGEXP_MATCH(pix_type, 'regex2') => 'foo', ...`)
2. The set of target kafka topics grows.  In this example there are 2, but if there were 5, they'd all have to fit in a long line of stellar.

I think it's not that stellar isn't equipped to handle it theoretically, it's just that for these kinds of use-cases, it's more readable to create something a bit more specific until we get multi-line Stellar available.  Ultimately, I consider this a stop-gap.

File: metron-platform/metron-parsers/README.md [hunk 2]
```diff
@@ -337,6 +337,28 @@ The following config will rename the fields `old_field` and `different_old_field
...
+```
+{
+...
+    "fieldTransformations" : [
+          {
+            "transformation" : "REGEX_SELECT"
+          , "input" : "my_input"
+          , "output" : "my_output"
+          , "config" : {
+            "awesome" : [ "^metron.*", "^mortron.*" ],
```
Comment (nickwallen) [inline metron-platform/metron-parsers/README.md]: A little nit.  It makes it harder to grok the docs when you use non-sensical phrases like 'boo', 'scary', etc in the primary example.  Is there an actual use case that we could show here instead?

Comment (nickwallen) [reply metron-platform/metron-parsers/README.md]: Same goes for the test cases, IMO, but I am not going to hold up the PR for that.

Comment (cestella) [reply metron-platform/metron-parsers/README.md]: I changed the parsers docs to reflect the use-case that I wrote up for the next PR.  That should fold in well together and be a real use-case.  Thanks for the feedback.

**Comment by ottobackwards** [comment]
" Ultimately, I consider this a stop-gap."

Yes.  What we are basically doing is writing a meta language on top of stellar.  In this case we are using that to make up for the meta language itself not supporting multi-line input ( and stellar's support for that as well )

A question would be, why would this transform not just be transformed into a match statement instead of adding a new function?

How much of this pain is from not using the UI for editing?  Would a ui editor that let you put in a multi line match make this pain better?

**Comment by cestella** [comment]
I think a decent UI that would make this a hell of a lot easier, but also not embedding stellar in JSON would help a lot.  Something like:
```
match {
  REGEXP_MATCH(pix_type, '^6-302.*') => 'cisco-6-302', 
  REGEXP_MATCH(pix_type, '^5-304.*') => 'cisco-5-304', 
  default => NULL
}
```
is a lot easier to read and deal with.  Also, enabling REGEXP_MATCH to take multiple regexes via a list in the 2nd arg would make things easier.

**Comment by nickwallen** [comment]
A big thumbs up to not embedding Stellar directly in JSON.  I think that would solve a number of problems.

**Comment by ottobackwards** [comment]
Why don't you create a jira for the REGEXP_MATCH

**Comment by cestella** [comment]
You got it [METRON-1655](https://issues.apache.org/jira/browse/METRON-1655)

**Pull request closed without merging**